src/database/utils.tsx

```diff
@@ -25,3 +26,141 @@ export const prepareFromStorage = (state: object): EpidemicModelPersistent => {
     { deep: true },
   );
 };
+
+type BatchOperationData = firebase.firestore.DocumentData;
+type BatchOperationRef = firebase.firestore.DocumentReference<
+  BatchOperationData
+>;
+type BatchDelete = {
+  type: "DELETE";
+  params: { ref: BatchOperationRef };
+};
+type BatchUpdate = {
+  type: "UPDATE";
+  params: { ref: BatchOperationRef; data: BatchOperationData };
+};
+type BatchSet = {
+  type: "SET";
+  params: {
+    ref: BatchOperationRef;
+    data: BatchOperationData;
+    options?: firebase.firestore.SetOptions;
+  };
+};
+type BatchOperation = BatchDelete | BatchSet | BatchUpdate;
+
+/**
+ * Wrapper around firebase.firestore.WriteBatch that abstracts away the batch size limit.
+ * Exposes methods and properties that would result in batch operations.
+ * Automatically chunks the batch when it's full to prevent overflows.
+ */
+export class BatchWriter {
+  private client: firebase.firestore.Firestore;
+  private currentBatch!: BatchOperation[];
+  // a single "write" can trigger additional operations (in effect,
+  // incrementing the batch size by >1); track them here to anticipate
+  // and prevent overflows
+  private currentBatchSize = 0;
+  // this limit is imposed by Firestore
+  readonly MAX_BATCH_SIZE = 500;
+  private operationBatches: BatchOperation[][] = [];
+  private pendingOpsCount = 0;
+
+  constructor(client: firebase.firestore.Firestore) {
+    this.client = client;
+    this.startBatch();
+  }
+
+  private startBatch() {
+    const newBatch: BatchOperation[] = [];
+    this.currentBatch = newBatch;
+    this.operationBatches.push(newBatch);
+  }
+
+  private trackWrite() {
+    this.currentBatchSize += 1 + this.pendingOpsCount;
```

Comment:
Could this not lead to double counting pending operations? For example, if I have 2 pending ops, and I pass through some code flow that leads to consecutive calls to `trackWrite`, wouldn't those pending ops get counted for each invocation of `trackWrite`? What's preventing that from happening?

I could also envision the answer being, "Yes, this is possible, but it's okay because all it would do is lead to us being more conservative in chunking up operations."

Reply:
they wouldn't get double-counted because `trackWrite` resets that counter to 0. But they could get counted out of order in that situation ... there is an expectation here that you will use this class by adding whatever server transforms you need and then immediately calling the "write" method that will trigger them. If you didn't follow that order — e.g. you queued up 500 objects that each rack up one or two `pendingOps` and then called `BatchWrite.update` 500 times — then `commit` would throw an error. 

It's an imperfect heuristic ... the truly officially Recommended way to do this is to make a Cloud Function and do more parallel writes instead of serial writes (I think the batching functionality is somewhat intentionally limited to encourage you to do this instead), but that seemed like it would be a bigger architectural change. I was kind of hoping we could get away with not doing it? But if you think this approach is too brittle I couldn't exactly argue with you ...

Reply:
Given prioritization and the relatively lowish likelihood of this drifting away from us, I think we can keep this as-is. But we should comment on `trackWrite` in big bold letters what you said here about the expected pattern.